### PR TITLE
fix: transaction value type to u256

### PR DESCRIPTION
**Pull request by lazyanubis**
Issues #4 send larger amounts of Eth
nonce: u64 => U256
gas_price: u64 => U256
gas_limit: u64 => U256
value: u64 => U256

### CODE REVIEW

File: e2e/backend/src/lib.rs [hunk 1]
```diff
@@ -69,9 +69,9 @@ async fn deploy_evm_contract(
         principal_id,
         bytecode,
         chain_id,
-        max_priority_fee_per_gas,
-        gas_limit,
-        max_fee_per_gas,
+        ic_evm_sign::u64_to_u256(max_priority_fee_per_gas),
```
Comment (LiveDuo) [inline e2e/backend/src/lib.rs:72]: I think it's better to fix the function's argument too to use U256.

File: e2e/backend/src/lib.rs [hunk 2]
```diff
@@ -94,11 +94,11 @@ async fn transfer_erc_20(
     let res = ic_evm_sign::transfer_erc_20(
         principal_id,
         chain_id,
-        max_priority_fee_per_gas,
-        gas_limit,
-        max_fee_per_gas,
+        ic_evm_sign::u64_to_u256(max_priority_fee_per_gas),
```
Comment (LiveDuo) [inline e2e/backend/src/lib.rs:97]: Same as above.

File: lib/src/lib.rs [hunk 3]
```diff
@@ -13,6 +13,7 @@ mod mocks;
 use mocks::{ic_call, ic_timestamp};
 
 mod utils;
+pub use utils::u64_to_u256;
```
Comment (LiveDuo) [inline lib/src/lib.rs:16]: Any reason why this import is `pub`?

File: lib/src/lib.rs [hunk 4]
```diff
@@ -183,11 +184,11 @@ pub async fn deploy_contract(
...
     }
 
-    let nonce: u64;
+    let nonce: U256;
     if let Some(user_transactions) = user.transactions.get(&chain_id) {
-        nonce = user_transactions.nonce;
+        nonce = utils::u64_to_u256(user_transactions.nonce);
     } else {
-        nonce = 0;
+        nonce = [0; 32];
```
Comment (LiveDuo) [inline lib/src/lib.rs:191]: Better to use [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero).

File: lib/src/lib.rs [hunk 5]
```diff
@@ -197,7 +198,7 @@ pub async fn deploy_contract(
         gas_limit,
         max_fee_per_gas,
         to: "0x".to_string(),
-        value: 0,
+        value: [0; 32],
```
Comment (LiveDuo) [inline lib/src/lib.rs:201]: 👉 [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero)

File: lib/src/lib.rs [hunk 6]
```diff
@@ -232,11 +233,11 @@ pub async fn transfer_erc_20(
...
     }
 
-    let nonce: u64;
+    let nonce: U256;
     if let Some(user_transactions) = user.transactions.get(&chain_id) {
-        nonce = user_transactions.nonce;
+        nonce = utils::u64_to_u256(user_transactions.nonce);
     } else {
-        nonce = 0;
+        nonce = [0; 32];
```
Comment (LiveDuo) [inline lib/src/lib.rs:240]: 👉 [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero)

File: lib/src/lib.rs [hunk 7]
```diff
@@ -248,7 +249,7 @@ pub async fn transfer_erc_20(
         gas_limit,
         max_fee_per_gas,
         to: contract_address,
-        value: 0,
+        value: [0; 32],
```
Comment (LiveDuo) [inline lib/src/lib.rs:252]: 👉 [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero)

File: lib/src/tests.rs [hunk 8]
```diff
@@ -22,11 +22,11 @@ fn sign_legacy_transaction() {
     let expected_address = "0x907dc4d0be5d691970cae886fcab34ed65a2cd66";
 
     let tx = transaction::TransactionLegacy {
-        nonce: 0,
-        gas_price: 0,
-        gas_limit: 0,
+        nonce: [0; 32],
```
Comment (LiveDuo) [inline lib/src/tests.rs:25]: Again [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero).

File: lib/src/tests.rs [hunk 9]
```diff
@@ -73,11 +73,11 @@ fn sign_eip2930_transaction() {
 
     let tx = transaction::Transaction2930 {
         chain_id: 1,
-        nonce: 0,
-        gas_price: 0,
-        gas_limit: 0,
+        nonce: [0; 32],
```
Comment (LiveDuo) [inline lib/src/tests.rs:76]: 👉 [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero)

File: lib/src/tests.rs [hunk 10]
```diff
@@ -124,12 +124,12 @@ fn sign_eip1559_transaction() {
 
     let tx = transaction::Transaction1559 {
         chain_id: 1,
-        nonce: 0,
-        max_priority_fee_per_gas: 0,
-        gas_limit: 0,
-        max_fee_per_gas: 0,
+        nonce: [0; 32],
```
Comment (LiveDuo) [inline lib/src/tests.rs:127]: 👉 [struct.U256.html#method.zero](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.zero)

File: lib/src/utils.rs [hunk 11]
```diff
@@ -22,7 +22,7 @@ pub fn get_address_from_public_key(public_key: Vec<u8>) -> Result<String, String
     Ok(address)
 }
 
-pub fn get_transfer_data(address: &str, amount: u64) -> Result<String, String> {
+pub fn get_transfer_data(address: &str, amount: [u8; 32]) -> Result<String, String> {
```
Comment (LiveDuo) [inline lib/src/utils.rs:25]: Can we use `U256` directly as an argument?

File: lib/src/utils.rs [hunk 12]
```diff
@@ -115,7 +139,8 @@ mod tests {
         let expected ="a9059cbb000000000000000000000000907dc4d0be5d691970cae886fcab34ed65a2cd660000000000000000000000000000000000000000000000000000000000000001";
 
         let address = "0x907dc4d0be5d691970cae886fcab34ed65a2cd66";
-        let amount = 1;
+        let mut amount = [0; 32];
```
Comment (LiveDuo) [inline lib/src/utils.rs:142]: We can use [struct.U256.html#method.from_dec_str](https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#method.from_dec_str) for more clarity.

File: lib/src/utils.rs [hunk 13]
```diff
@@ -124,7 +149,8 @@ mod tests {
     fn get_transfer_data_with_invalid_address() {
         let expected = Err("Invalid address".to_string());
         let address = "0x00";
-        let value = 1;
+        let mut value = [0; 32];
```
Comment (LiveDuo) [inline lib/src/utils.rs:152]: Same as above.

File: lib/src/transaction.rs [hunk 14]
```diff
@@ -14,17 +17,19 @@ pub trait Sign {
...
 }
 
+pub type U256 = [u8; 32];
+
 pub struct TransactionLegacy {
     pub chain_id: u64,
-    pub nonce: u64,
-    pub gas_price: u64,
-    pub gas_limit: u64,
+    pub nonce: U256,
```
Comment (LiveDuo) [inline lib/src/transaction.rs:28]: I think we should follow geth more closely. This is the mistake we did and missed the correct `amount` when we first developed the library. IMO if we follow geth there's less chance for a bug similar to https://github.com/nikolas-con/ic-evm-sign/issues/4.

```rs
pub chain_id: U256, // Not part of legacy but for compatibility with the other tx types
pub nonce: u64,
pub gas_price: U256,
pub gas_limit: u64,
pub value: U256,
```

For reference:
https://github.com/ethereum/go-ethereum/blob/master/core/types/tx_legacy.go#L27

File: lib/src/transaction.rs [hunk 15]
```diff
@@ -195,18 +200,18 @@ impl Sign for TransactionLegacy {
...
         Ok(self.nonce)
     }
 }
 
 pub struct Transaction2930 {
     pub chain_id: u64,
-    pub nonce: u64,
-    pub gas_price: u64,
-    pub gas_limit: u64,
+    pub nonce: U256,
```
Comment (LiveDuo) [inline lib/src/transaction.rs:210]: Should follow geth too (similar to legacy tx).

For reference:
https://github.com/ethereum/go-ethereum/blob/master/core/types/transaction.go#L72

File: lib/src/transaction.rs [hunk 16]
```diff
@@ -392,19 +397,19 @@ impl Sign for Transaction2930 {
...
     }
 }
 
 pub struct Transaction1559 {
     pub chain_id: u64,
-    pub nonce: u64,
-    pub max_priority_fee_per_gas: u64,
-    pub gas_limit: u64,
-    pub max_fee_per_gas: u64,
+    pub nonce: U256,
```
Comment (LiveDuo) [inline lib/src/transaction.rs:407]: Should follow geth too (similar to legacy tx).

For reference:
https://github.com/ethereum/go-ethereum/blob/master/core/types/transaction.go#L72

**Review by LiveDuo — COMMENTED** [review]
Thanks for the PR.

In summary there are a few comments:
1. Use the U256 library traits for more clarity
2. Follow Geth transaction types
3. Fix e2e canister function parameters

For more info see exact review comments.

File: lib/src/lib.rs (hunk 3 above)
Comment (lazyanubis) [reply lib/src/lib.rs:16]: Maybe some one wants to use it

**Pull request closed without merging**